### PR TITLE
Permutation example

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test_data/asm/generic_permutation.asm [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    first * (z - 1) = 0;
+
+    // Update rule:
+    // z' = z * (beta - a) / (beta - b)
+    (beta - left_compressed) * z' = z * (beta - right_compressed);
+};
+
+let compress_selected_exprs: int, expr, expr, expr, expr[] -> expr = |id, alpha, beta, sel, exprs|
+    sel * (compress_columns(alpha, exprs) + id + beta - 1) + 1;
```
Comment (georgwiese) [inline test_data/asm/generic_permutation.asm:27]: Where does the ID come from? We don't really need it, do we? I think it's a way to share the accumulator between several permutations, which we're not doing here?

File: test_data/asm/generic_permutation.asm [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    // Update rule:
+    // z' = z * (beta - a) / (beta - b)
+    (beta - left_compressed) * z' = z * (beta - right_compressed);
+    // we need to return something...
+    []
+};
+
+let compress_selected_exprs: expr, expr, expr, expr[] -> expr = |alpha, beta, sel, exprs|
+    sel * (compress_columns(alpha, exprs) + beta - 1) + 1;
```
Comment (georgwiese) [inline test_data/asm/generic_permutation.asm:30]:
```suggestion
    sel * (beta - compress_columns(alpha, exprs) - 1) + 1;
```
I think both is sound, but this makes it clear we're evaluating $$\prod_{i = 1}^n (X - a_i)$$ at a random point.

**Comment by chriseth** [comment]
Wondering if this PR is superseded, but it adds the `new_witness_in_stage` builtin. That's all it does.

**Comment by leonardoalt** [comment]
is it worth it to revive it?

**Comment by leonardoalt** [comment]
I'm closing this for now, assuming it has been superseded.

**Pull request closed without merging**